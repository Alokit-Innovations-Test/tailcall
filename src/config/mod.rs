mod config;
mod from_document;
pub mod group_by;
mod into_document;
mod key_values;
mod n_plus_one;
mod server;
mod source;
mod js_plugin;

pub use config::*;
pub use key_values::*;
pub use server::*;
pub use source::*;
pub use js_plugin::*;

fn is_default<T: Default + Eq>(val: &T) -> bool {
  *val == T::default()
}
