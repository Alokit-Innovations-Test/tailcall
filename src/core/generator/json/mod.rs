mod field_base_url_generator;
mod http_directive_generator;
mod mutation_generator;
mod query_generator;
mod schema_generator;
mod types_generator;
mod url_utils;

pub use field_base_url_generator::FieldBaseUrlGenerator;
pub use mutation_generator::OperationTypeGenerator;
pub use schema_generator::SchemaGenerator;
pub use types_generator::GraphQLTypesGenerator;
