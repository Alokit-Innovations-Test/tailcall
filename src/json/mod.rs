mod async_value;

mod group_by;
mod json_like;
mod json_schema;
mod serde_value;
pub use group_by::*;
pub use json_like::*;
pub use json_schema::*;
