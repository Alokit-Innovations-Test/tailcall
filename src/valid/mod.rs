mod cause;
mod error;
mod valid;
mod zipped;

pub use cause::*;
pub use error::*;
pub use valid::*;
pub use zipped::*;
