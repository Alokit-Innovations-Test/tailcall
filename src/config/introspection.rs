use std::collections::BTreeMap;

use async_graphql::futures_util::future::join_all;
use reqwest;
use serde_json::json;

// GraphQL introspection response types.
#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionResult {
  pub data: Option<IntrospectionData>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionData {
  pub __schema: IntrospectionSchema,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionSchema {
  #[serde(rename = "queryType")]
  pub query_type: IntrospectionType,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct IntrospectionType {
  pub fields: Vec<Field>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Field {
  pub name: String,
  pub args: Option<Vec<Arg>>,
  #[serde(rename = "type")]
  pub type_: Option<Type_>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Arg {
  pub name: String,
  #[serde(rename = "type")]
  pub type_: Type_,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
pub struct Type_ {
  pub name: Option<String>,
  pub kind: Option<String>,
  #[serde(rename = "ofType")]
  pub of_type: Option<Box<Type_>>,
  pub fields: Option<Vec<Field>>,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum TypeKind {
  Scalar,
  Object,
  Interface,
  Union,
  Enum,
  InputObject,
  List,
  NonNull,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, Default)]
pub struct IntrospectionResults(pub BTreeMap<String, IntrospectionResult>);

impl IntrospectionResults {
  pub fn merge_right(mut self, other: Self) -> Self {
    let mut merged = self.0;
    for (key, value) in other.0 {
      merged.insert(key, value);
    }
    self.0 = merged;
    self
  }
}

pub async fn introspect_endpoint(graphql_url: &String) -> Result<IntrospectionResult, Box<dyn std::error::Error>> {
  let introspection_query: String =
  json!({"query": "query { __schema { queryType { name fields { name args { name type { name kind ofType {name }}} type { name kind ofType { name fields {name} } fields { name }}  } } } }"}).to_string();

  let result = reqwest::Client::new()
    .post(graphql_url)
    .header("Content-Type", "application/json")
    .body(introspection_query)
    .send()
    .await?
    .json::<IntrospectionResult>()
    .await?;

  Ok(result)
}

pub async fn introspect_endpoints(graphql_urls: Vec<String>) -> IntrospectionResults {
  let mut results = BTreeMap::new();
  let introspect_futures = graphql_urls
    .iter()
    .map(|url| async move { (url.clone(), introspect_endpoint(url).await.unwrap()) });
  let joined = join_all(introspect_futures).await;
  for (url, introspection_result) in joined {
    results.insert(url.clone(), introspection_result.clone());
  }
  IntrospectionResults(results)
}

pub fn get_arg_type(
  introspection_result: &IntrospectionResult,
  query_name: &String,
  arg_name: &String,
) -> Option<String> {
  introspection_result
    .data
    .as_ref()
    .and_then(|data| {
      data
        .__schema
        .query_type
        .fields
        .iter()
        .find(|field| field.name == *query_name)
    })
    .and_then(|field| field.args.as_ref())
    .and_then(|args| args.iter().find(|arg| arg.name == *arg_name))
    .and_then(|arg| {
      arg.type_.name.as_ref().map_or_else(
        || {
          let kind = arg.type_.kind.clone().unwrap_or_default();
          match &arg.type_.of_type {
            Some(type_) => type_.name.as_ref().map(|name| format_name(name, &kind)),
            None => None,
          }
        },
        |name| Some(name.clone()),
      )
    })
}

fn format_name(name: &String, kind: &String) -> String {
  if kind == "LIST" {
    format!("[{}]", name)
  } else if kind == "NON_NULL" {
    format!("{}!", name)
  } else {
    name.clone()
  }
}

#[cfg(test)]
mod tests {
  use std::fs;
  use std::path::PathBuf;

  use httpmock::Method::POST;
  use httpmock::MockServer;

  use crate::config::introspection::*;

  fn load_mock_introspection_result() -> String {
    let mut mock_result_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    mock_result_path.push("tests/data/introspection-result.json");
    fs::read_to_string(mock_result_path).unwrap()
  }

  fn mock_introspection_results() -> BTreeMap<String, IntrospectionResult> {
    let result: IntrospectionResult = serde_json::from_str(&load_mock_introspection_result()).unwrap();
    let mut introspection_result: BTreeMap<String, IntrospectionResult> = BTreeMap::new();
    introspection_result.insert("http://localhost:8000/graphql".to_string(), result);
    introspection_result
  }

  #[test]
  fn test_get_arg_type() {
    let introspection_results = mock_introspection_results();
    let introspection_result = introspection_results
      .get(&"http://localhost:8000/graphql".to_string())
      .unwrap();
    assert_eq!(
      get_arg_type(introspection_result, &"user".to_string(), &"id".to_string()).unwrap(),
      "Int"
    );
    assert_eq!(
      get_arg_type(introspection_result, &"post".to_string(), &"id".to_string()).unwrap(),
      "Int!"
    );
    assert_eq!(
      get_arg_type(introspection_result, &"post".to_string(), &"doesntexist".to_string()),
      None
    );
    assert_eq!(
      get_arg_type(introspection_result, &"doesntexist".to_string(), &"id".to_string()),
      None
    );
  }

  #[tokio::test]
  async fn test_introspect_endpoint() {
    let server = MockServer::start();
    let result = load_mock_introspection_result();

    server.mock(|when, then| {
      when.method(POST).path("/graphql");
      then.status(200).header("content-type", "application/json").body(result);
    });

    let graphql_url = server.url("/graphql");

    let introspect_result = introspect_endpoint(&graphql_url).await;
    assert_eq!(
      introspect_result
        .unwrap()
        .data
        .unwrap()
        .__schema
        .query_type
        .fields
        .len(),
      2
    );
  }
}
