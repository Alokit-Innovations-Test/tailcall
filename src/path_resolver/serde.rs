use super::PathResolver;
use crate::json::JsonLike;

impl PathResolver for serde_json::Value {
    fn get_path_value<Path>(&self, path: &[Path]) -> Option<async_graphql::Value>
    where
        Path: AsRef<str>,
    {
        // FIXME: drop this implementation
        self.get_path(path)
            .and_then(|v| async_graphql::Value::from_json(v.to_owned()).ok())
    }
}
