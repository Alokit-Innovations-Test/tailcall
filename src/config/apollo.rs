use anyhow::Context;
use serde::{Deserialize, Serialize};

use crate::config::ConfigReaderContext;
use crate::mustache::Mustache;

#[derive(Serialize, Deserialize, Clone, Debug, Default, PartialEq, Eq, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Apollo {
    ///
    /// Setting `apiKey` for Apollo.
    pub api_key: String,
    ///
    /// Setting `graphRef` for Apollo in the format <graphId>@<variant>.
    pub graph_ref: String,
    ///
    /// Setting `userVersion` for Apollo.
    #[serde(default = "default_user_version")]
    pub user_version: String,
    ///
    /// Setting `platform` for Apollo.
    #[serde(default = "default_platform")]
    pub platform: String,
    ///
    /// Setting `version` for Apollo.
    #[serde(default = "default_version")]
    pub version: String,
}

fn default_user_version() -> String {
    "1.0".to_string()
}

fn default_platform() -> String {
    "platform".to_string()
}

fn default_version() -> String {
    "1.0".to_string()
}

impl Apollo {
    pub fn render_mustache(&mut self, reader_ctx: &ConfigReaderContext) -> anyhow::Result<()> {
        let Apollo { api_key, graph_ref, user_version, platform, version } = self;

        let api_key_tmpl = Mustache::parse(api_key)?;
        *api_key = api_key_tmpl
            .render_string(reader_ctx)
            .context("apiKey is not defined")?;

        let graph_ref_tmpl = Mustache::parse(graph_ref)?;
        *graph_ref = graph_ref_tmpl
            .render_string(reader_ctx)
            .context("graphRef is not defined")?;

        let user_version_tmpl = Mustache::parse(user_version)?;
        if let Some(rendered) = user_version_tmpl.render_string(reader_ctx) {
            *user_version = rendered;
        }

        let platform_tmpl = Mustache::parse(platform)?;
        if let Some(rendered) = platform_tmpl.render_string(reader_ctx) {
            *platform = rendered;
        }

        let version_tmpl = Mustache::parse(version)?;
        if let Some(rendered) = version_tmpl.render_string(reader_ctx) {
            *version = rendered;
        }

        Ok(())
    }
}
