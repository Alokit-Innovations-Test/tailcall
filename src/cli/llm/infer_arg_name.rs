use genai::chat::{ChatMessage, ChatRequest, ChatResponse};
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use serde_json::json;

use super::{Error, Result, Wizard};
use crate::core::config::transformer::{ArgumentInfo, RenameArgs};
use crate::core::config::{Config, Resolver};
use crate::core::valid::{Valid, Validator};
use crate::core::{AsyncTransform, Mustache, Transform};

pub struct InferArgName {
    wizard: Wizard<Question, Answer>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct MetaData {
    name: String,
    #[serde(rename = "outputType")]
    output_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct OperationDefinition {
    argument: MetaData,
    field: MetaData,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Answer {
    suggestions: Vec<String>,
}

impl TryFrom<ChatResponse> for Answer {
    type Error = Error;

    fn try_from(response: ChatResponse) -> Result<Self> {
        let message_content = response.content.ok_or(Error::EmptyResponse)?;
        let text_content = message_content.text_as_str().ok_or(Error::EmptyResponse)?;
        Ok(serde_json::from_str(text_content)?)
    }
}

#[derive(Clone, Serialize)]
struct Question {
    definition: OperationDefinition,
}

impl TryInto<ChatRequest> for Question {
    type Error = Error;

    fn try_into(self) -> Result<ChatRequest> {
        let input2 = OperationDefinition {
            argument: {
                MetaData {
                    name: "input2".to_string(),
                    output_type: "Article".to_string(),
                }
            },
            field: {
                MetaData {
                    name: "createPost".to_string(),
                    output_type: "Post".to_string(),
                }
            },
        };

        let input = serde_json::to_string_pretty(&Question { definition: input2 })?;
        let output = serde_json::to_string_pretty(&Answer {
            suggestions: vec![
                "createPostInput".into(),
                "postInput".into(),
                "articleInput".into(),
                "noteInput".into(),
                "messageInput".into(),
            ],
        })?;

        let template_str = include_str!("prompts/infer_arg_name.md");
        let template = Mustache::parse(template_str);

        let context = json!({
            "input": input,
            "output": output,
            "count": 5,
        });

        let rendered_prompt = template.render(&context);
        let request = ChatRequest::new(vec![
            ChatMessage::system(rendered_prompt),
            ChatMessage::user(serde_json::to_string(&self.definition)?),
        ]);
        Ok(request)
    }
}

impl InferArgName {
    pub fn new(model: String, secret: Option<String>) -> InferArgName {
        Self { wizard: Wizard::new(model, secret) }
    }

    pub async fn generate(&self, config: &Config) -> Result<IndexMap<String, ArgumentInfo>> {
        let mut mapping: IndexMap<String, ArgumentInfo> = IndexMap::new();

        for (type_name, type_) in config.types.iter() {
            // collect all the args that's needs to be processed with LLM.
            for (field_name, field) in type_.fields.iter() {
                if field.args.is_empty() {
                    continue;
                }
                // filter out query params as we shouldn't change the names of query params.
                for (arg_name, arg) in field.args.iter().filter(|(k, _)| match &field.resolver {
                    Some(Resolver::Http(http)) => !http.query.iter().any(|q| &q.key == *k),
                    _ => true,
                }) {
                    let question = OperationDefinition {
                        argument: MetaData {
                            name: arg_name.to_string(),
                            output_type: arg.type_of.name().to_owned(),
                        },
                        field: MetaData {
                            name: field_name.to_string(),
                            output_type: field.type_of.name().to_owned(),
                        },
                    };

                    let question = Question { definition: question };

                    let mut delay = 3;
                    loop {
                        let answer = self.wizard.ask(question.clone()).await;
                        match answer {
                            Ok(answer) => {
                                tracing::info!(
                                    "Suggestions for Argument {}: [{:?}]",
                                    arg_name,
                                    answer.suggestions,
                                );
                                mapping.insert(
                                    arg_name.to_owned(),
                                    ArgumentInfo::new(
                                        answer.suggestions,
                                        field_name.to_owned(),
                                        type_name.to_owned(),
                                    ),
                                );
                                break;
                            }
                            Err(e) => {
                                // TODO: log errors after certain number of retries.
                                if let Error::GenAI(_) = e {
                                    // TODO: retry only when it's required.
                                    tracing::warn!(
                                        "Unable to retrieve a name for the type '{}'. Retrying in {}s",
                                        type_name,
                                        delay
                                    );
                                    tokio::time::sleep(tokio::time::Duration::from_secs(delay))
                                        .await;
                                    delay *= std::cmp::min(delay * 2, 60);
                                }
                            }
                        }
                    }
                }
            }
        }

        Ok(mapping)
    }
}

impl AsyncTransform for InferArgName {
    type Value = Config;
    type Error = Error;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> crate::core::valid::Valid<Self::Value, Self::Error> {
        match self.generate(&value).await {
            Ok(suggested_names) => {
                match RenameArgs::new(suggested_names)
                    .transform(value)
                    .to_result()
                {
                    Ok(v) => Valid::succeed(v),
                    Err(e) => Valid::fail(Error::Err(e.to_string())),
                }
            }
            Err(err) => Valid::fail(err),
        }
    }
}

#[cfg(test)]
mod test {
    use genai::chat::{ChatRequest, ChatResponse, MessageContent};

    use super::{Answer, Question};
    use crate::cli::llm::infer_arg_name::{OperationDefinition, MetaData};

    #[test]
    fn test_to_chat_request_conversion() {
        let question = Question {
            definition: OperationDefinition {
                argument: MetaData {
                    name: "input2".to_string(),
                    output_type: "Article".to_string(),
                },
                field: MetaData {
                    name: "createPost".to_string(),
                    output_type: "Post".to_string(),
                },
            },
        };
        let request: ChatRequest = question.try_into().unwrap();
        insta::assert_debug_snapshot!(request);
    }

    #[test]
    fn test_chat_response_parse() {
        let resp = ChatResponse {
            content: Some(MessageContent::Text(
                "{\"suggestions\":[\"createPostInput\",\"postInput\",\"articleInput\",\"noteInput\",\"messageInput\"]}".to_owned(),
            )),
            ..Default::default()
        };
        let answer = Answer::try_from(resp).unwrap();
        insta::assert_debug_snapshot!(answer);
    }
}
