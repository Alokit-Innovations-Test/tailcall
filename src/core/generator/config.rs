use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
