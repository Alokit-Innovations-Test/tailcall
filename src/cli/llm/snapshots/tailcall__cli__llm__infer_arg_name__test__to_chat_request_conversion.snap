---
source: src/cli/llm/infer_arg_name.rs
expression: request
---
ChatRequest {
    system: None,
    messages: [
        ChatMessage {
            role: System,
            content: Text(
                "Given the Operation Definition of GraphQL, suggest 5 meaningful names for the argument names.\nThe name should be concise and preferably a single word.\n\nExample Input:\n{\n  \"definition\": {\n    \"argument\": {\n      \"name\": \"input2\",\n      \"outputType\": \"Article\"\n    },\n    \"field\": {\n      \"name\": \"createPost\",\n      \"outputType\": \"Post\"\n    }\n  }\n}\n\nExample Output:\n{\n  \"suggestions\": [\n    \"createPostInput\",\n    \"postInput\",\n    \"articleInput\",\n    \"noteInput\",\n    \"messageInput\"\n  ]\n}\n\nEnsure the output is in valid JSON format.\n",
            ),
            extra: None,
        },
        ChatMessage {
            role: User,
            content: Text(
                "{\"argument\":{\"name\":\"input2\",\"outputType\":\"Article\"},\"field\":{\"name\":\"createPost\",\"outputType\":\"Post\"}}",
            ),
            extra: None,
        },
    ],
}
